### ISSUE TITLE
SyncWorker: setLogins - Tried to use a null handle 

### ISSUE THREAD

**Issue by vladikoff** [mozilla-mobile/android-components#5715]
Sometimes the Fenix app gets into a Handle invalid state. In my experience I think it was when I worked with local servers with servers sending errors.

Logcat reports the following:

 
Handle is invalid: Tried to use a null handle (this object has probably been closed)

```
Work [ id=113443b2-a824-415e-a1cf-11f55b0383b9, tags={ mozilla.components.service.fxa.sync.WorkManagerSyncWorker, Debounce, Common } ] failed because it threw an exception/error
                         E  java.util.concurrent.ExecutionException: mozilla.appservices.syncmanager.UnexpectedError: Handle is invalid: Tried to use a null handle (this object has probably been closed)
                         E      at androidx.work.impl.utils.futures.AbstractFuture.getDoneValue(AbstractFuture.java:516)
                         E      at androidx.work.impl.utils.futures.AbstractFuture.get(AbstractFuture.java:475)
                         E      at androidx.work.impl.WorkerWrapper$2.run(WorkerWrapper.java:284)
                         E      at androidx.work.impl.utils.SerialExecutor$Task.run(SerialExecutor.java:91)
                         E      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
                         E      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
                         E      at java.lang.Thread.run(Thread.java:919)
                         E  Caused by: mozilla.appservices.syncmanager.UnexpectedError: Handle is invalid: Tried to use a null handle (this object has probably been closed)
                         E      at mozilla.appservices.syncmanager.RustError.intoException(RustError.kt:40)
                         E      at mozilla.appservices.syncmanager.SyncManager.setLogins(SyncManager.kt:96)
                         E      at mozilla.components.service.fxa.sync.WorkManagerSyncWorker.doSync(WorkManagerSyncManager.kt:342)
                         E      at mozilla.components.service.fxa.sync.WorkManagerSyncWorker$doWork$2.invokeSuspend(WorkManagerSyncManager.kt:326)
                         E      at mozilla.components.service.fxa.sync.WorkManagerSyncWorker$doWork$2.invoke(Unknown Source:10)
                         E      at mozilla.components.service.sync.logins.SyncableLoginsStore.unlocked(AsyncLoginsStorage.kt:466)
                         E      at mozilla.components.service.sync.logins.SyncableLoginsStore$unlocked$1.invokeSuspend(Unknown Source:12)
                         E      at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                         E      at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
                         E      at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:561)
                         E      at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:727)
                         E      at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:667)
                         E      at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:655)
                         I  Worker result FAILURE for Work [ id=113443b2-a824-415e-a1cf-11f55b0383b9, tags={ mozilla.components.service.fxa.sync.WorkManagerSyncWorker, Debounce, Common } ]

```

This happens when there are 401 errors from the sync server

**Comment by grigoryk** [comment]
`Tried to use a null handle` means that `raw` handle value inside of `DatabaseLoginsStorage` was `0`.

`raw` starts off as `0`, is set to an internal map handle when we "unlock" the db, and is reset back to 0 when it's locked.

So this stacktrace means one of two things:
- we're reading the handle after the db was unlocked, and then locked
- db was never unlocked

In the background sync worker, "password storage" is accessed via a global reference set on `GlobalSyncableStoreProvider`. In Fenix, that refers to an instance of `syncablePasswordsStorage` in Core.kt, the same object instance that's used by the rest of the application. It's locked/unlocked during regular browser operation - during browsing when interacting with login forms, when accessing password manager UI, etc.

Background sync worker needs to keep the password storage unlocked for the duration of the logins sync, which in practice means "for the duration of the whole sync", since we don't have a more granular visibility via the places' sync manager. This is a rather large time window, and we currently have no guarantees that the handle we've obtained either - won't be `0` when the worker starts up, or won't become invalid sometime during sync.

I see two obvious solutions to this:
1) "unlock" the password storage during app startup, and keep it unlocked
2) or, maintain a ref count when we request storage layer to be unlock/lock, making sure that it's only "truly" locked when the ref count reaches 0 (e.g. there are no consumers who care about storage being unlocked).

(2) just seems like a good thing to do regardless, since we essentially have a global resource which is accessed concurrently by various parts of the browser, and we need to make sure things behave correctly regardless how the store is accessed.

More to come...

**Comment by grigoryk** [comment]
Seems to me like the answer is "do both". There's a lot of overhead in unlocking the store - we need to obtain an encryption key from the system keystore (which I think could be _very_ expensive, and device-dependent due to presence/absence of secure hardware enclaves), establish a new sqlite connection, set some pragmas on it, etc. We're also not "locking" the keystore in a sense that user interaction (like device unlock) is required to access key material inside. So with those two things in mind, it seems like keeping password store unlocked while our process is around is no worse than the constant lock/unlock dance we currently have, but it will be easier to reason about, and significantly improve perf of any operation that requires password storage.

**Comment by grigoryk** [comment]
To complete the loop on the perf impact - here's a plot of how long `unlock` operations take (I don't think this includes overhead of accessing the keystore) - https://sql.telemetry.mozilla.org/queries/68211/source#172407. It's rather slow in the fast cases, and _very_ slow (on the order of seconds) at the tail end.

**Comment by st3fan** [comment]
@grigoryk @ekager Is this similar to https://github.com/mozilla-mobile/android-components/issues/6044 - that one has a very similar stack trace.

**Comment by grigoryk** [comment]
@st3fan yup, this is the root cause.

**Comment by ebalazs-sv** [comment]
Verified as fixed on Beta Migration Build #3 with OnePlus 5T (Android 9) and Fenix does not crash when testing the scenarios:
- accessing Saved Logins for the second time while performing a Sync;
- going to accounts.firefox.com;
- going to Saved logins;
- typing/deleting an email address in a login field on Facebook/Vimeo;
- performing a search on Vimeo.com.

I will move this to Done. Thanks!
